### ISSUE TITLE
Introduce lifecycle hooks for models

### ISSUE THREAD

**Issue by mweststrate** [mobxjs/mobx-state-tree#76]
suggestions:
- beforeDispose
- afterCreate
- beforeApplySnapshot
- afterApplySnapshot
- afterCreateSnapshot
- beforeChange (like intercept)

..But only if there are good use cased :-D

**Comment by mattiamanzati** [comment]
+1

afterCreateInstance to setup reactions! :D

**Comment by mweststrate** [comment]
Proposal:

* `preprocessSnapshot(snapshot) => snapshot` (called before applying a snapshot, both during create and later on)
* `postprocessSnapshot(snapshot) => snapshot` (called when creating a snapshot)
* `didCreate()` called after instantation, but not during reconcilation
* `didAttach()` called when node gets attached to a tree (not called if attached before creation)
* `willDetach()` called before detaching from a tree
* `willDestroy()` called before a node is killed

**Comment by RainerAtSpirit** [comment]
Naming convention before/after resonates better with me.

**Comment by mattiamanzati** [comment]
@mweststrate Maybe add a preprocess and postprocess for patches too?

**Comment by mweststrate** [comment]
Merged by #98 and released as 0.3.3.

Following lifecycle hooks are now available:
* `afterCreate`
* `beforeDestroy`
* `afterAttach`
* `beforeDetach`

More lifecycle hooks should be demand / use case driven. E.g. processing snapshots.

Also introduced the convenience method `addDisposer` which is useful in `afterCreate`, avoid the need to setup a `beforeDestroy` just to fire disposers. 

E.g.:

```
const Todo = types.model({
   afterCreate() {
       addDisposer(this, reaction(() => getSnapshot(this), s => submitSnapshotToServer(s)))
   }
})
```
